Add docs for updating external dependencies (#1280)
python/devguide: developer-workflow/sbom.rst

   that correct package is removed from the SBOM.
5. Commit the changes to :cpy-file:`Misc/sbom.spdx.json` and
   :cpy-file:`Tools/build/generate_sbom.py`.

Updating external dependencies (cpython-source-deps)
----------------------------------------------------

> Dependencies for Windows CPython builds are `stored in a separate repository <https://github.com/python/cpython-source-deps>`_

Reviewer: Some binaries are also stored in https://github.com/python/cpython-bin-deps, though generally they should also have sources in the source-deps repo. Is this distinction important here?

Author: Do any of the cpython-bin-deps get shipped along with the CPython artifacts? If they're derived from the cpython-source-deps repository I think we should be okay.

Reviewer: I think the only one that isn't derived from `cpython-source-deps` is `vcruntime140.dll`, which comes from our repo to make sure we always get the latest one and not whichever GHA build machine we're on.

and then fetched during builds of CPython for Windows in the script :cpy-file:`PCbuild/get_externals.bat`.

> In this script the libraries to fetch are designated by ``{name}-{version}`` git refs being added to the ``libraries`` variable.
> SBOM tooling in the CPython repository matches these git refs in order to build the :cpy-file:`Misc/externals.spdx.json`
> SBOM file. When updating external dependencies for a CPython branch:

Reviewer 2: ```suggestion
In this script the libraries to fetch are designated by ``{name}-{version}`` Git refs being added to the ``libraries`` variable.
SBOM tooling in the CPython repository matches these Git refs in order to build the :cpy-file:`Misc/externals.spdx.json`
SBOM file.

When updating external dependencies for a CPython branch:
```

> 1. Push the update to the ``cpython-source-deps`` repository and create a new git tag.

Reviewer 2: ```suggestion
1. Push the update to the ``cpython-source-deps`` repository and create a new Git tag.
```

> 1. Push the update to the ``cpython-source-deps`` repository and create a new git tag.

Reviewer: Maybe worth noting that this can only be done by a core committer, and we don't accept PRs for it (because we need to verify the sources have come from the right source and are unmodified, and our trust boundary for this is "has the commit bit").

Also might be worth noting that sometimes there's a build step involved and the core committer will then push a tag to `cpython-bin-deps` that will actually be used in the build. Tcl/Tk, libffi and OpenSSL are all in this group.

In practice for contributors, what this usually means is that they should post an issue requesting the updated version, wait for a core dev to say the tags are ready, and then the contributor can continue with the following steps.

Author: I've addressed this comment in https://github.com/python/devguide/pull/1280/commits/b32b691f14434299aa29b054061a139d77dea8f0. Do you think we should cover the cpython-bin-deps part here as well?

Reviewer: Not in the same note, but it ought to be documented somewhere. At the very least, we should mention the `cpython-bin-deps` repo at least once so that someone reading this knows to look there.

2. Update the entry for the project in ``get_externals.bat``.

> 3. Run ``make regen-sbom`` in the CPython source repository.

Reviewer 2: Can the people who will likely update Windows dependencies run `make` commands, or should we also give Windows-specific instructions?

Reviewer: I would prefer Windows specific instructions.

I believe `PCbuild\build.bat --regen` ought to do it, if that was set up correctly, but I haven't tried it.

Author: That hasn't been setup, I didn't realize there were multiple regen flows. I [created an issue](https://github.com/python/cpython/issues/116122) and assigned myself, will submit a PR.

Author: I added the command in https://github.com/python/devguide/pull/1280/commits/cefc01d48aaa142585ddde2cef77e66755a02da3

> 4. Verify the metadata (like version, download location) in ``externals.spdx.json`` SBOM is updated as expected with ``git diff``

Reviewer 2: ```suggestion
4. Verify the metadata (like version, download location) in ``externals.spdx.json`` SBOM is updated as expected with ``git diff``.
```

   that correct package is removed from the SBOM.
5. Commit the changes to :cpy-file:`Misc/sbom.spdx.json` and
   :cpy-file:`Tools/build/generate_sbom.py`.

> Updating external dependencies (cpython-source-deps)
> ----------------------------------------------------

Reviewer 3: ```suggestion
Updating external dependencies (``cpython-source-deps``)
--------------------------------------------------------
```
Should this get some markup too?

> Dependencies for Windows CPython builds are `stored in a separate repository <https://github.com/python/cpython-source-deps>`_
> and then fetched during builds of CPython for Windows in the script :cpy-file:`PCbuild/get_externals.bat`.

Reviewer 3: ```suggestion
Dependencies for Windows CPython builds are `stored in a separate repository
<https://github.com/python/cpython-source-deps>`_ and then fetched during
builds of CPython for Windows in the script :cpy-file:`PCbuild/get_externals.bat`.
```
Rewrap to keep the lines short.

> In this script the libraries to fetch are designated by ``{name}-{version}`` Git refs being added to the ``libraries`` variable.
> SBOM tooling in the CPython repository matches these Git refs in order to build the :cpy-file:`Misc/externals.spdx.json`
> SBOM file.

Reviewer 3: ```suggestion
In this script the libraries to fetch are designated by ``{name}-{version}``
Git refs being added to the ``libraries`` variable.
SBOM tooling in the CPython repository matches these Git refs in order
to build the :cpy-file:`Misc/externals.spdx.json` SBOM file.
```
Rewrap.

When updating external dependencies for a CPython branch:

> 1. Push the update to the ``cpython-source-deps`` repository and create a new Git tag.
> 2. Update the entry for the project in ``get_externals.bat``.
> 3. Run ``make regen-sbom`` or ``PCbuild/build.bat --regen`` in the CPython source repository.
> 4. Verify the metadata (like version, download location) in ``externals.spdx.json`` SBOM is updated as expected with ``git diff``.
> 5. Commit the changes and have them merged together.

Reviewer 3: ```suggestion
1. Push the update to the ``cpython-source-deps`` repository and
   create a new Git tag.
2. Update the entry for the project in ``get_externals.bat``.
3. Run ``make regen-sbom`` or ``PCbuild/build.bat --regen``
   in the CPython source repository.
4. Use ``git diff`` to verify that the metadata (like version, download location)
   in ``externals.spdx.json`` SBOM is updated as expected.
5. Commit the changes and have them merged together.
```
Rewrap and rephrase 4.

   that correct package is removed from the SBOM.
5. Commit the changes to :cpy-file:`Misc/sbom.spdx.json` and
   :cpy-file:`Tools/build/generate_sbom.py`.

> Updating external dependencies (``cpython-source-deps``)

Reviewer: It's a little bit awkward because the _intention_ is that it's not, but in practice, no other platform has needed it yet. (Maybe that's the clarification that is needed?)

--------------------------------------------------------

 .. note::
   Only core developers can push to the ``cpython-source-deps`` repository.
   Pull requests are not accepted because the sources must be verified for
   integrity. Contributors should create an issue requesting the updated
   version and then wait for a core developer to prepare the new version
   before proceeding with the next steps below.

> Dependencies for Windows CPython builds are `stored in a separate repository

Reviewer 4: This would be resolved by better clarity at the start of this section.

<https://github.com/python/cpython-source-deps>`_ and then fetched during
builds of CPython for Windows in the script :cpy-file:`PCbuild/get_externals.bat`.

In this script the libraries to fetch are designated by ``{name}-{version}``

> Git refs being added to the ``libraries`` variable.

Reviewer 4: This is still unclear.
